src/rest/server/router.go

```diff
@@ -456,26 +456,32 @@ func withMiddleware(h http.Handler, name string) http.Handler {
 // authentication and authorization. This middleware will return
 // 401 response if authentication fails and 403 if authorization
 // fails.
-func authMiddleware(inner http.Handler) http.Handler {
+func authMiddleware(inner http.Handler, auth UserAuth) http.Handler {
+	if !auth.Any() {
+		return inner
+	}
+
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		rc, r := GetContext(r)
+		rc.ClientAuth = auth
+		glog.Infof("Valid Auth Modes: %s", rc.ClientAuth)
```

Comment:
>Also, old global ClientAuth variable had an advantage -- it was possible to change the auth mode dynamically, without service resatrt. Thats not possible now, unless you change UserAuth parameter everywhere to pointers.

Maps are always passed by reference. You can still change the UserAuth by accessing the UserAuth from the request context and changing it, or by changing the original reference you passed to it.

Also, I prototyped the other method, of providing a different auth middleware as well, but I found that instead you had a lot of duplicated code so I found this method preferably, but  am open to it. I have a branch with those changes as well.

Reply:
Didnt notice UserAuth is a map. So, dynamic mode changes are still possible.

Regarding the other prototype -- I felt current approach had touches lot of APIs and function calls. In some places it looks like a hack. Having separate middleware for CLI sounds lot cleaner approach to me.. Since we are nearing the release let us pick a version which is tested well.

Reply:
Sure, the problem is the withMiddleware function is used in many places, so all of these function that call it would need to be changed anyway... Either way you will end up touching a lot of functions. 

Another idea is I can add the UserAuth reference to the Router type to facilitate modifying it after the fact.

Reply:
Actually NewRouter does not return the router object, it returns a handler, so forget that.

Reply:
Also, just to verify, I tried modifying the clientAuth mode after creating the router and starting the server and it works. I started with jwt and cert mode, then set password mode and verified that password works and the logs show jwt,cer and password are enabled.